File: js/rosetta.js
// === Core Data Patterns Rosetta (JS) ===

// iteration + access
characters.forEach(c => c.name);
for (let i = characters.length - 1; i >= 0; i--) characters[i];

// map / filter / reduce (manual + native)
const map = (a,f)=>a.reduce((r,x,i)=>(r.push(f(x,i)),r),[]);
const filter = (a,f)=>a.reduce((r,x,i)=>(f(x,i)&&r.push(x),r),[]);
const reduce = (a,f,i)=>a.reduce(f,i);

// transform
const names = characters.map(c=>c.name);
const chaotic = characters.filter(c=>c.chaos>5);
const totalChaos = characters.reduce((s,c)=>s+c.chaos,0);

// non-mutation update (shallow)
const addFriend = (c,f)=>({...c,friends:[...c.friends,f]});

// search
characters.find(c=>c.name==="Daffy Duck");
characters.findIndex(c=>c.chaos>=10);

// sorting + slicing
[...characters].sort((a,b)=>b.chaos-a.chaos).slice(0,3);

// nested traversal
characters.flatMap(c=>c.episodes.map(e=>({...e,who:c.name})));

// flatten (recursive)
const flatten = a =>
  a.reduce((r,v)=>r.concat(Array.isArray(v)?flatten(v):v),[]);

// sets + maps
const uniqueFriends = new Set(characters.flatMap(c=>c.friends));
const speciesMap = new Map(characters.map(c=>[c.name,c.species]));

// indexing for O(1) lookup
const episodeIndex = characters.flatMap(c=>c.episodes)
  .reduce((m,e)=>(m[e.title]=e,m),{});

// safe access
characters[0]?.episodes?.[0]?.likes ?? 0;

// recursion over unknown shape
const upper = x =>
  typeof x==="string" ? x.toUpperCase() :
  Array.isArray(x) ? x.map(upper) :
  x && typeof x==="object"
    ? Object.fromEntries(Object.entries(x).map(([k,v])=>[k,upper(v)]))
    : x;

// generators (lazy)
function* chaosLevels(){
  for(const c of characters) yield c.chaos;
}

// async iteration (sketch)
async function* episodeTitles(){
  for(const c of characters)
    for(const e of c.episodes)
      yield await Promise.resolve(e.title);
}

File: py/rosetta.py
# === Core Data Patterns Rosetta (Py) ===

# iteration + access
[c["name"] for c in characters]
for i in range(len(characters) - 1, -1, -1): characters[i]

# map / filter / reduce (manual + native)
def my_map(lst, fn): return [fn(x, i) for i, x in enumerate(lst)]
def my_filter(lst, fn): return [x for i, x in enumerate(lst) if fn(x, i)]
def my_reduce(lst, fn, acc):
    for i, x in enumerate(lst): acc = fn(acc, x, i)
    return acc

# transform
names = [c["name"] for c in characters]
chaotic = [c for c in characters if c["chaos"] > 5]
total_chaos = sum(c["chaos"] for c in characters)

# non-mutation update (shallow)
def add_friend(c, f):
    return {**c, "friends": [*c["friends"], f]}

# search
next((c for c in characters if c["name"] == "Daffy Duck"), None)
next((i for i,c in enumerate(characters) if c["chaos"] >= 10), -1)

# sorting + slicing
top3 = sorted(characters, key=lambda c: c["chaos"], reverse=True)[:3]

# nested traversal / flatten one level
episodes = [
    {**e, "who": c["name"]}
    for c in characters for e in c["episodes"]
]

# flatten (recursive)
def flatten(x):
    return [y for i in x for y in (flatten(i) if isinstance(i,list) else [i])]

# sets + dicts
unique_friends = {f for c in characters for f in c["friends"]}
species_map = {c["name"]: c["species"] for c in characters}

# indexing for O(1) lookup
episode_index = {
    e["title"]: e
    for c in characters for e in c["episodes"]
}

# safe access
likes = characters[0].get("episodes",[{}])[0].get("likes", 0)

# recursion over unknown shape
def upper(x):
    if isinstance(x,str): return x.upper()
    if isinstance(x,list): return [upper(v) for v in x]
    if isinstance(x,dict): return {k:upper(v) for k,v in x.items()}
    return x

# generators (lazy)
def chaos_levels():
    for c in characters: yield c["chaos"]

# async iteration (sketch)
import asyncio
async def episode_titles():
    for c in characters:
        for e in c["episodes"]:
            await asyncio.sleep(0)
            yield e["title"]
